### ISSUE TITLE
Optional in PA2

### ISSUE THREAD

**Issue by SweetPotato0213** [nus-cs2030/2021-s2#214]
May I know if we have to use Optional in PA2?

**Comment by chamath2030** [comment]
no it's not compulsory 

lastly, shoutout to my GitHubbers Gwajoon, Wentinggy, Keithlimhs, naozitios, kaiyuuu, itsjiaming and elton. Continue living it big in the realm of SoC, hope to see you guys in other modules in the future.

**Comment by brendancjz** [comment]
Hi there, it's not necessary but it is helpful because we can use its .map() method.

**Comment by e0543492** [comment]
You dont have to, but you will find with much easier, especially when you are handling null values

**Comment by rendao1010** [comment]
Hi there, it is not compulsory to use Optionals! Hope this helps :-)

**Comment by arnav0401** [comment]
From my experience, I believe it is not compulsory to use Optionals! Hope this helps :)
Have a good day and good luck for finals

**Comment by lozhaowei** [comment]
Hi! I believe it is not necessary to use Optionals in this PA, but perhaps you can try it out and see if it is any different from your implementation!

**Comment by Martini-Chwang** [comment]
Hello! Yepp although it is good practice to use Optionals, it is not compulsory to use it if you've used null values during the PE.

**Comment by dumbasssoul** [comment]
yep i asked the TA and he said it was okay! i guess you can just use mapper.apply(sth) instead of .map(mapper)

**Comment by marvinljw** [comment]
Yeapp It is not necessary to use Optional in this PA. I think it will be better if you follow your own implementation

**Comment by mltnsia** [comment]
It is not necessary but certainly more convienient!

**Comment by karintang** [comment]
Is not necessary to use Optional but it will make ur life easy.

**Comment by i-sack** [comment]
Hi, it is not necessary! My initial method did not involve Optionals, so it limited me to be unable to use the .map() and .flatMap() methods. However, it can also be done without Optionals (aka null), by slightly tweaking your outputs.

**Comment by oliverguicw** [comment]
Hi that's a great question! I tried to do without using Optionals as well but in the end I think using optionals is better design and the code is a lot shorter as well.

**Comment by ARPspoofing** [comment]
Hi actually I tried not doing optionals and it made my code super messy. So I would recommend using Optionals.

**Comment by jcng2308** [comment]
> Hi, it is not necessary! My initial method did not involve Optionals, so it limited me to be unable to use the .map() and .flatMap() methods. However, it can also be done without Optionals (aka null), by slightly tweaking your outputs.

Hello there would it be possible to see your code snippet how you did your map and flatmap without Optionals

**Comment by jeromesjj08** [comment]
> > Hi, it is not necessary! My initial method did not involve Optionals, so it limited me to be unable to use the .map() and .flatMap() methods. However, it can also be done without Optionals (aka null), by slightly tweaking your outputs.
> 
> Hello there would it be possible to see your code snippet how you did your map and flatmap without Optionals

I did mine without using using Optionals also. (Cause I am super bad at that topic)
To replace the .map prof used, you can use mapper.apply(). (It is shown in one of his recitation video) 
For Flatmap is slightly more tricky. What I did was that I stored the function in a dummy Aggregate<S,T> to make sure the function gets stored somewhere since it came from a static method. Hope this helps, all the best =)

**Comment by jcng2308** [comment]
> > > Hi, it is not necessary! My initial method did not involve Optionals, so it limited me to be unable to use the .map() and .flatMap() methods. However, it can also be done without Optionals (aka null), by slightly tweaking your outputs.
> > 
> > 
> > Hello there would it be possible to see your code snippet how you did your map and flatmap without Optionals
> 
> I did mine without using using Optionals also. (Cause I am super bad at that topic)
> To replace the .map prof used, you can use mapper.apply(). (It is shown in one of his recitation video)
> For Flatmap is slightly more tricky. What I did was that I stored the function in a dummy Aggregate<S,T> to make sure the function gets stored somewhere since it came from a static method. Hope this helps, all the best =)

Thank you for your reply, i managed to figure out the dummy Aggregate<S, T>

**Comment by chence08** [comment]
it's a painful lesson indeed, to not have used Optionals early on... ><

**Comment by gableejh** [comment]
i managed to do it without optionals too!

**Comment by chelleadel** [comment]
Hello! I agree with everyone else! I finished it without optional and find optional more complicated!

**Comment by keithcharlesc** [comment]
Hi as what the rest have mentioned it is not necessary but as what have been suggested, the handling of null might be a more complicated task!

**Comment by Ryanng-zr** [comment]
Hi, it was not necessary! I suggest using null so that you can make minimal changes to your PA2. It worked well for me!

**Comment by hua-fong** [comment]
It is not necessary!

**Comment by ayangler** [comment]
I agree with the rest that optionals are not necessary! Having to repeatedly check for null values made my code rather messy though.

**Comment by marcuslim835** [comment]
i tried with and without optionals and i realised that less changes are made using optionals for me, maybe you could do the same to see which saves you more code?

**Comment by saiyara1512** [comment]
I originally used null, but found it hard to continue in the map and flatmap method, so ended up using Optional (which was less tedious for me). Good luck!

**Comment by Hanyu0118** [comment]
Using Optional is more concise. Even though i did not use optional before modification, i still added Optional in my final code. :((

**Comment by ydymegan** [comment]
Optional makes things a lot easier but not strictly necessary

**Comment by sangyouan** [comment]
ultimately, it is better practice to use optional, and depending on how much you had progressed for your code without optional, it might be quicker to make your code optional instead.

**Comment by Martini-Chwang** [comment]
Using Optionals instead of null is good practice, so might as well try using it if possible, even though it's not compulsory in this case

**Comment by evewzhiru** [comment]
its not compulsory to use optionals. I did mine without using optionals

**Comment by Joeltan15** [comment]
Abit late but its not compulsory to use Optionals.  Basically the aggregate class will contain three items, a T object1, a S object2 and a saved function.  Because it is now possible to create a class Aggregate with simply a function, this would lead to the T and S objects being null.

So there are two alternatives: 
a.) check if the objects is null
b.) set the objects as optional and check later

either method A or B works for this lab.  The case in which you would definitely have to use method B is if you have a statement somewhere which later assumes that the objects1/2 are not null, and you will use methods belonging to objects1/2.  This would lead to nullpointerexceptions during runtime.
